js/coinsbit.js

```diff
@@ -0,0 +1,1114 @@
+        //         {
+        //             id: 99601,
+        //             type: "buy",
+        //             time: 1640108002.455,
+        //             amount: "323144.05252658",
+        //             price: "0.008",
+        //             total: "2585.15242021"
+        //         },
+        //
+        //    /history/result
+        //         {
+        //             tid: 99601,
+        //             type: "buy",
+        //             date: 1640108002,
+        //             amount: "323144.05252658",
+        //             price: "0.008",
+        //             total: "2585.15242021"
+        //         },
+        //
+        //      /account/trades :
+        //      [
+        //         {
+        //           time: '1640258718.402',
+        //           fee: '0.01572657',
+        //           price: '3931.64271609',
+        //           amount: '0.002',
+        //           id: '1595921',
+        //           dealOrderId: '8631729321',
+        //           role: '2',
+        //           deal: '7.86328543'
+        //         },
+        //         ...
+        //       ]
+        const symbol = this.safeValue (market, 'symbol');
+        const id = this.safeString2 (trade, 'id', 'tid');
+        const orderId = this.safeString (trade, 'dealOrderId');
+        const timestamp = this.safeTimestamp2 (trade, 'time', 'date');
+        const price = this.safeString (trade, 'price');
+        const amount = this.safeString (trade, 'amount');
+        const side = this.safeString (trade, 'type');
+        const cost = this.safeString (trade, 'deal');
+        const fee = this.safeString (trade, 'fee');
+        let takerOrMaker = '';
+        if ('role' in trade) {
+            takerOrMaker = this.safeString (trade, 'role') === '2' ? 'taker' : 'maker';
+        } else {
+            takerOrMaker = 'taker';
+        }
+        return this.safeTrade ({
+            'info': trade,
+            'timestamp': timestamp,
+            'datetime': this.iso8601 (timestamp),
+            'symbol': symbol,
+            'id': id,
+            'order': orderId,
+            'type': undefined,
+            'takerOrMaker': takerOrMaker,
+            'side': side,
+            'price': price,
+            'amount': amount,
+            'cost': cost,
+            'fee': fee,
+        }, market);
+    }
+
+    async createOrder (symbol, type, side, amount, price = undefined, params = {}) {
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        const request = {
+            'market': market['id'],
+        };
+        if (!price) {
+            throw new InvalidOrder (this.id + ' createOrder needs price field');
+        }
+        request['side'] = side;
+        request['amount'] = this.amountToPrecision (symbol, amount);
+        request['price'] = this.priceToPrecision (symbol, price);
+        const response = await this.privatePostOrderNew (this.extend (request, params));
```

Comment:
Let's rewrite the above lines 645-657:

```JavaScript
    async createOrder (symbol, type, side, amount, price = undefined, params = {}) {
        if (type !== 'limit') {
            throw new InvalidOrder (this.id + ' createOrder supports limit orders only');
        }
        await this.loadMarkets ();
        const market = this.market (symbol);
        const request = {
            'market': market['id'],
            'side': side,
            'amount': this.amountToPrecision (symbol, amount),
            'price': this.priceToPrecision (symbol, price),
        };
        const response = await this.privatePostOrderNew (this.extend (request, params));
```

Reply:
done, but I've question. maybe you have hints - the exchange supports Market Orders in UI, but couldnt do it from API. Has there been such occasions with other exchanges?

Reply:
> Has there been such occasions with other exchanges?

Yes, that is pretty common with the low-end exchanges.
Also, we could test if setting the `price` to `0` actually places a market order.
And we can test if they will accept an order without a price.

Reply:
yes, I had them tested before commented, and neither 'unsetting' price field, and neither 0 worked there. So, I assumed Market Orders not supported from API. However, I will ask their support through ticket (though my experience with asking API questions through support to even major exchange support reps, is quite bad).

Reply:
If they don't support market orders, then we should throw early on `(type !== 'limit')`.